### PR TITLE
added dom obj counting to decide sequential/parallel layout (#10110)

### LINKED ISSUE #10110: Dynamically switch to sequential layout if only a small amount of reflow needs to happen
Parallel layout overhead is overkill if only a small number of nodes need to be reflowed. In the extreme case, if one chain of nodes from parent to child needs to be reflowed, then parallel layout cannot help at all since it's sequentialized, and all of the traversal code is pure overhead!

The easiest way to solve this, I think, would be to have an atomic counter of the number of nodes that were discovered to be dirty (`incremental.rs` can compute this, perhaps) and have the layout thread consult that number. If below a threshold (8? wild guess) then parallel layout will not be spawned. Note that we don't want this to regress performance on pages that have a small number of nodes but lots of reflow to do: consider four paragraphs of 10,000 words each, for instance, which has a small number of nodes but is the ideal case for parallel layout. So we'll probably need a couple of heuristics here and tune appropriately.

cc @paulrouget

### PR DESCRIPTION
**Pull request by notriddle**
This is a rebased version of #11713

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #10110 (github issue number if applicable).
- [X] There are no tests for these changes because it's an optimization with no visible behavioral changes

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12862)

### CODE REVIEW

**Comment by emilio** [comment]
This was already reviewed in the original PR, so r=me,metajack I guess, though I think a check with browserhtml would be nice, to see if we parallelize in those cases.

@metajack what do you think about the following comments (excluding nits).

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/layout_thread/lib.rs, line 403 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA3gf7sY8GGCa19EDN:-KPA3gf7sY8GGCa19EDO:b-doqxq4) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/layout_thread/lib.rs#L403)):*
> ```Rust
>                                     thread_state::LAYOUT,
>                                     layout_threads);
> ```

nit: extra newline.

---

*[components/layout_thread/lib.rs, line 741 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA3mie1ugobGUoIxiM:-KPA3mifC9JCZ8YmpmWg:b-a3xnq3) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/layout_thread/lib.rs#L741)):*
> ```Rust
> 
>         // ... as do each of the LayoutWorkers, if present.
>         let ref traversal = self.parallel_traversal;
> ```

maybe merge these two lines? `self.parallel_traversal.heap_size_of_tls()`

---

*[components/layout_thread/lib.rs, line 802 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA7nUw3WvLltBmi3bx:-KPA7nUw3WvLltBmi3by:bqn19wo) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/layout_thread/lib.rs#L802)):*
> ```Rust
>     /// crash.
>     fn exit_now(&mut self) {
>         let ref mut traversal = self.parallel_traversal;
> ```

nit: self.parallel_traversal.shutdown()?

---

*[components/layout_thread/lib.rs, line 1458 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA46-kV5o0EqBHQXtF:-KPA46-lwNcEJ3nvytxa:bae5l9i) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/layout_thread/lib.rs#L1458)):*
> ```Rust
>                     let profiler_metadata = self.profiler_metadata();
> 
>                     debug!("layout: post style parallel? {}", self.parallel_flag);
> ```

Is this really needed?

---

*[components/script/dom/document.rs, line 415 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA4sMNloN6e0kaLB3b:-KPA4sMNloN6e0kaLB3c:bwhm5zj) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/dom/document.rs#L415)):*
> ```Rust
> 
>     pub fn increment_dom_count(&self) {
>         self.dom_count.set(self.dom_count.get() + 1);
> ```

Also, can you file a followup to update this when nodes are removed or inserted dynamically? Right now it seems to me that `dom_count` is something like `dom_elements_at_first_page_parse`, and it doesn't seem really nice not taking into account elements dinamically created. In fact, I doubt this will even turn parallel layout on for bhtml, for example, whose index.html page is almost empty.

---

*[components/script/dom/servohtmlparser.rs, line 284 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA4ShSrUl5U3A2C-wD:-KPA4ShSrUl5U3A2C-wE:b-doqxq4) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/dom/servohtmlparser.rs#L284)):*
> ```Rust
>         };
> ```

nit: extra newline.

---

*[components/script/dom/servohtmlparser.rs, line 342 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA4ZgvVrXRMg_qthoI:-KPA4ZgvVrXRMg_qthoJ:btfj0ke) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/dom/servohtmlparser.rs#L342)):*
> ```Rust
>         &self.pending_input
>     }
> ```

nit: So is this, and you can probably take rid of the other below.

---

*[components/script/parse/html.rs, line 94 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA7-GDPg6q0prSnhNo:-KPA7-GEmSLLPe2h6N6j:b-575ny5) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/parse/html.rs#L94)):*
> ```Rust
>         }
> 
>         self.document.increment_dom_count();
> ```

Should this also go in `insert()`? where text nodes are created?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862#-:-KPA4kEV09pLQun7Vhw1:b-4az3hc)*

**Comment by notriddle** [comment]
*[components/layout_thread/lib.rs, line 403 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA3gf7sY8GGCa19EDN:-KPB4vUNrxWZz8K3g-F7:b-896fix) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/layout_thread/lib.rs#L403)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> nit: extra newline.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by notriddle** [comment]
*[components/layout_thread/lib.rs, line 741 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA3mie1ugobGUoIxiM:-KPB4vzKQ3HQqh5Y27Rl:b-896fix) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/layout_thread/lib.rs#L741)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> maybe merge these two lines? `self.parallel_traversal.heap_size_of_tls()`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by notriddle** [comment]
*[components/layout_thread/lib.rs, line 802 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA7nUw3WvLltBmi3bx:-KPB4wqVf2eDKPLEM5G6:b-896fix) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/layout_thread/lib.rs#L802)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> nit: self.parallel_traversal.shutdown()?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by notriddle** [comment]
*[components/layout_thread/lib.rs, line 1458 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA46-kV5o0EqBHQXtF:-KPB4xQTacsNdvUOhiG0:b-896fix) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/layout_thread/lib.rs#L1458)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> Is this really needed?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by notriddle** [comment]
*[components/script/dom/servohtmlparser.rs, line 342 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA4ZgvVrXRMg_qthoI:-KPB5WR0s_QNcVl3tKsZ:b-896fix) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/dom/servohtmlparser.rs#L342)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> nit: So is this, and you can probably take rid of the other below.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by notriddle** [comment]
*[components/script/dom/servohtmlparser.rs, line 284 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA4ShSrUl5U3A2C-wD:-KPB5Xh6FzUfyh22W4mD:b-896fix) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/dom/servohtmlparser.rs#L284)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> nit: extra newline.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by notriddle** [comment]
*[components/script/parse/html.rs, line 94 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA7-GDPg6q0prSnhNo:-KPB5xDyYVp2XN8oJ_3r:b-hufnyc) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/parse/html.rs#L94)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> Should this also go in `insert()`? where text nodes are created?
</details>
Text nodes can't be laid out in parallel, so a page with a bunch of text nodes but very little structure probably wouldn't benefit from parallel layout.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by emilio** [comment]
Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/script/parse/html.rs, line 94 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA7-GDPg6q0prSnhNo:-KPB6nqyg3MJoJJaaPi8:b-q2d3tt) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/parse/html.rs#L94)):*
<details><summary><i>Previously, notriddle (Michael Howell) wrote…</i></summary>
> Text nodes can't be laid out in parallel, so a page with a bunch of text nodes but very little structure probably wouldn't benefit from parallel layout.
</details>
They can be styled in parallel though, and styling is a decent amount of the whole layout process.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by notriddle** [comment]
*[components/script/dom/document.rs, line 415 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA4sMNloN6e0kaLB3b:-KPB7o_nQ0Kp-BO2GVxw:b-cgi3d) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/dom/document.rs#L415)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> Also, can you file a followup to update this when nodes are removed or inserted dynamically? Right now it seems to me that `dom_count` is something like `dom_elements_at_first_page_parse`, and it doesn't seem really nice not taking into account elements dinamically created. In fact, I doubt this will even turn parallel layout on for bhtml, for example, whose index.html page is almost empty.
</details>
Done. #12871

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by notriddle** [comment]
Review status: 4 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/parse/html.rs, line 94 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12862#-KPA7-GDPg6q0prSnhNo:-KPBAPDL2-P3KDhnXb-G:b-896fix) ([raw file](https://github.com/servo/servo/blob/8cfdb9c26ac1af102097cede41c68d7e00519947/components/script/parse/html.rs#L94)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> They can be styled in parallel though, and styling is a decent amount of the whole layout process.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862)*

**Comment by nox** [comment]
I think the counting should be a new `weight` field on `Node` and be updated on every DOM mutation. We already need to notify all ancestors on mutations that something in the tree changed, and I have a few ideas to use that weight as an optimisation. For example we know for sure that a node with a lesser weight than an other node cannot be one of its ancestors.

**Comment by notriddle** [comment]
In this PR, or a follow-up?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12862#-:-KPFtwe5Y2x-UgnuEwa_:bo0mjpe)*

**Comment by nox** [comment]
@notriddle Yeah not blocking this.

**Comment by emilio** [comment]
r=me if you think testing with bhtml is not necessary, or if you have done it and it parallelizes/it's not a severe perf impact.

**Comment by notriddle** [comment]
Now doing it based in the DOM node code, so it'll trigger in any complex-enough page.

**Comment by emilio** [comment]
So we've found a few issues with styling today that may explain the above failures, see https://bugzilla.mozilla.org/show_bug.cgi?id=1300505

**Comment by jdm** [comment]
So is this blocked on work being done elsewhere? Where is that being tracked?

**Comment by notriddle** [comment]
Yeah, the issue @emilio pointed out.

**Comment by emilio** [comment]
Yes, that's tracked in that gecko bug, and should be over once #13202 lands, since it removes that cache.

**Comment by notriddle** [comment]
#13202 landed.

@bors-servo try

**Comment by emilio** [comment]
Glad I debug assertions are working on the buildbots :)

These are somewhat scary, hopefully a minor bug.

```
attempt to subtract with overflow (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at /home/servo/buildbot/slave/linux-rel-wpt/build/components/script/dom/document.rs:446)
  │ stack backtrace:
  │    0:     0x7f033521752d - backtrace::backtrace::trace::h16372ee7bf1517e5
  │    1:     0x7f0335217bc2 - backtrace::capture::Backtrace::new::hcc43c50c4b11c693
  │    2:     0x7f0334caf0ad - servo::main::_{{closure}}::h8e9c5971861a4720
  │    3:     0x7f03368e8ee4 - std::panicking::rust_panic_with_hook
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:452
  │    4:     0x7f03368e8da4 - std::panicking::begin_panic<collections::string::String>
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:413
  │    5:     0x7f03368e8cc9 - std::panicking::begin_panic_fmt
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:397
  │    6:     0x7f03368e8c57 - std::panicking::rust_begin_panic
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:373
  │    7:     0x7f03369256bd - core::panicking::panic_fmt
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libcore/panicking.rs:69
  │    8:     0x7f03369255f4 - core::panicking::panic
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libcore/panicking.rs:49
  │    9:     0x7f033565d499 - script::dom::node::Node::remove::hffed4b4d7b4fc127
  │   10:     0x7f033565ab1c - script::dom::node::Node::insert::h2c144011ff38f76a
  │   11:     0x7f033565c3a2 - script::dom::node::Node::replace_all::h04920a0df1ad2b70
```

**Comment by notriddle** [comment]
Apparently, the DOM observer stuff runs synchronously. Incrementing before I run those callbacks should work.

**Comment by emilio** [comment]
Cool part, all tests are green, bad part, tidy fails :)

**Comment by notriddle** [comment]
@bors-servo try

\*sigh\*

**Comment by emilio** [comment]
The layout pieces look good to me, but I'm not sure the `browsing_context` check is the one that should be done... So I'd prefer @jdm or @Ms2ger to sign-off that.

**Comment by notriddle** [comment]
I chose the browsing context because if there's no browsing context, there's no layout, so this optimization is inapplicable.

**Comment by Ms2ger** [comment]
Also, rather than referencing #10110, write up some in-tree docs of what we're doing here and why, and point there.

**Comment by notriddle** [comment]
The root document element doesn't get bound to the tree (being the root), but it can be removed from the tree. So let's start with a number for it. (`document.removeChild(document.documentElement)` is the thing all the failing tests had in common)

@bors-servo try

**Comment by notriddle** [comment]
wat

**Comment by notriddle** [comment]
This has to be a race condition, but I don't understand how it could be racey! The DOM runs in a single thread, and the tests all pass when I run it "locally" (actually in Janitor).

**Comment by notriddle** [comment]
Let's see if that fixed it.

**Comment by notriddle** [comment]
[I've finally managed to pass all the DOM tests](http://build.servo.org/builders/linux-rel-wpt/builds/1304). That fixed 'em.

```
  ▶ FAIL [expected PASS] /css21_dev/html4/abs-pos-non-replaced-vlr-149.htm
  └   → /css21_dev/html4/abs-pos-non-replaced-vlr-149.htm 495c94c5bdb9351da4864d1dffcb6741003f6a42
/css21_dev/html4/reference/abs-pos-non-replaced-vlr-009-ref.htm 760d37f04422ccc35f8b05dda55d0a8c5e6b2a3b
Testing 495c94c5bdb9351da4864d1dffcb6741003f6a42 == 760d37f04422ccc35f8b05dda55d0a8c5e6b2a3b

  ▶ Unexpected subtest result in /css-transitions-1_dev/html/changing-while-transition.htm:
  └ PASS [expected FAIL] changing transition-property / values
```

Now for more slogging through reftests that fail in sequential mode.

**Comment by notriddle** [comment]
[It passes all the CSS tests on Mac, apparently](http://build.servo.org/builders/mac-rel-css/builds/4605/steps/test/logs/stdio)

File: components/script/dom/htmltitleelement.rs [hunk 1]
```diff
@@ -71,9 +71,12 @@ impl VirtualMethods for HTMLTitleElement {
         }
     }
 
-    fn bind_to_tree(&self, is_in_doc: bool) {
+    fn bind_to_tree(&self, tree_in_doc: bool) {
+        if let Some(ref s) = self.super_type() {
```
Comment (emilio) [inline components/script/dom/htmltitleelement.rs]: You can unwrap here, right? Also above.

**Review by emilio — COMMENTED** [review]
Those are known intermittents:

 * https://github.com/servo/servo/issues/14478
 * https://github.com/servo/servo/issues/14425

r=me if tests pass and with the comment below addressed if you feel it's necessary :)

**Comment by notriddle** [comment]
That's not what the vast majority of the existing ones do.

@bors-servo r=emilio

**Pull request merged**